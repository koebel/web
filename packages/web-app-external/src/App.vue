<template>
  <iframe
    v-if="appUrl && method === 'GET'"
    :src="appUrl"
    class="oc-width-1-1 oc-height-1-1"
    :title="iFrameTitle"
    allowfullscreen
  />
  <div v-if="appUrl && method === 'POST' && formParameters" class="oc-height-1-1">
    <form :action="appUrl" target="app-iframe" method="post">
      <input ref="subm" type="submit" :value="formParameters" class="oc-hidden" />
      <div v-for="(item, key, index) in formParameters" :key="index">
        <input :name="key" :value="item" type="hidden" />
      </div>
    </form>
    <iframe
      name="app-iframe"
      class="oc-width-1-1 oc-height-1-1"
      :title="iFrameTitle"
      allowfullscreen
    />
  </div>
</template>

<script lang="ts">
import { stringify } from 'qs'
import { PropType, computed, defineComponent, unref, nextTick, ref, watch, VNodeRef } from 'vue'
import { useTask } from 'vue-concurrency'
import { useGettext } from 'vue3-gettext'

import { Resource } from 'web-client/src'
import { urlJoin } from 'web-client/src/utils'
import { queryItemAsString, useRequest, useRouteQuery, useStore } from 'web-pkg/src/composables'
import { configurationManager } from 'web-pkg/src/configuration'

export default defineComponent({
  name: 'ExternalApp',
  props: {
    resource: { type: Object as PropType<Resource>, required: true }
  },
  emits: ['update:applicationName'],
  setup(props, { emit }) {
    const language = useGettext()
    const store = useStore()

    const { $gettext } = language
    const { makeRequest } = useRequest()

    const appNameQuery = useRouteQuery('app')
    const appUrl = ref()
    const formParameters = ref({})
    const method = ref()
    const subm: VNodeRef = ref()

    const capabilities = computed(() => store.getters['capabilities'])
    const applicationName = computed(() => {
      const appName = queryItemAsString(unref(appNameQuery))
      emit('update:applicationName', appName)
      return appName
    })

    const iFrameTitle = computed(() => {
      return $gettext('"%{appName}" app content area', {
        appName: unref(applicationName)
      })
    })

    const errorPopup = (error) => {
      store.dispatch('showErrorMessage', {
        title: $gettext('An error occurred'),
        desc: error,
        error
      })
    }

    const loadAppUrl = useTask(function* () {
      try {
        const fileId = props.resource.fileId
        const baseUrl = urlJoin(
          configurationManager.serverUrl,
          unref(capabilities).files.app_providers[0].open_url
        )

        const query = stringify({
          file_id: fileId,
          lang: language.current,
          ...(unref(applicationName) && { app_name: unref(applicationName) })
        })

        const url = `${baseUrl}?${query}`
        const response = yield makeRequest('POST', url, {
          validateStatus: () => true
        })

        if (response.status !== 200) {
          switch (response.status) {
            case 425:
              errorPopup(
                $gettext(
                  'This file is currently being processed and is not yet available for use. Please try again shortly.'
                )
              )
              break
            default:
              errorPopup(response.data?.message)
          }

          const error = new Error('Error fetching app information')
          throw error
        }

        if (!response.data.app_url || !response.data.method) {
          const error = new Error('Error in app server response')
          throw error
        }

        appUrl.value = response.data.app_url
        method.value = response.data.method

        if (response.data.form_parameters) {
          formParameters.value = response.data.form_parameters
        }

        if (method.value === 'POST' && formParameters.value) {
          // eslint-disable-next-line vue/valid-next-tick
          yield nextTick()
          unref(subm).click()
        }
      } catch (e) {
        console.error('web-app-external error', e)
        throw e
      }
    }).restartable()

    watch(
      props.resource,
      () => {
        loadAppUrl.perform()
      },
      { immediate: true }
    )

    return {
      appUrl,
      formParameters,
      iFrameTitle,
      method,
      subm
    }
  }
})
</script>
