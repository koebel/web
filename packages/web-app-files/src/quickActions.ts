import { createQuicklink } from './helpers/share'
import { eventBus } from 'web-pkg/src/services/eventBus'
import { SideBarEventTopics } from 'web-pkg/src/composables/sideBar'
import { Resource } from 'web-client'
import { Language } from 'vue3-gettext'
import { ClientService, PasswordPolicyService } from 'web-pkg'
import { Ability } from 'web-client/src/helpers/resource/types'
import { Store } from 'vuex'

export function canShare(item, store) {
  const { capabilities } = store.state.user
  if (!capabilities.files_sharing || !capabilities.files_sharing.api_enabled) {
    return false
  }
  if (item.isReceivedShare() && !capabilities.files_sharing.resharing) {
    return false
  }
  return item.canShare()
}

export function showQuickLinkPasswordModal({ $gettext, store, passwordPolicyService }, onConfirm) {
  const modal = {
    variation: 'passive',
    title: $gettext('Set password'),
    cancelText: $gettext('Cancel'),
    confirmText: $gettext('Set'),
    hasInput: true,
    inputDescription: $gettext('Passwords for links are required.'),
    inputPasswordPolicy: passwordPolicyService.getPolicy(),
    inputGeneratePasswordMethod: () => passwordPolicyService.generatePassword(),
    inputLabel: $gettext('Password'),
    inputType: 'password',
    onInput: () => store.dispatch('setModalInputErrorMessage', ''),
    onPasswordChallengeCompleted: () => store.dispatch('setModalConfirmButtonDisabled', false),
    onPasswordChallengeFailed: () => store.dispatch('setModalConfirmButtonDisabled', true),
    onCancel: () => store.dispatch('hideModal'),
    onConfirm: async (password) => {
      onConfirm(password)
    }
  }

  return store.dispatch('createModal', modal)
}

interface QuickLinkContext {
  ability: Ability
  clientService: ClientService
  item: Resource
  language: Language
  store: Store<any>
  passwordPolicyService: PasswordPolicyService
}

export default {
  collaborators: {
    id: 'collaborators',
    label: ($gettext) => $gettext('Add people'),
    icon: 'user-add',
    iconFillType: undefined,
    handler: () => eventBus.publish(SideBarEventTopics.openWithPanel, 'sharing#peopleShares'),
    displayed: canShare
  },
  quicklink: {
    id: 'quicklink',
    label: ($gettext) => $gettext('Copy link'),
    icon: 'link',
    iconFillType: undefined,
    handler: async ({
      ability,
      clientService,
      item,
      language,
      store,
      passwordPolicyService
    }: QuickLinkContext) => {
      const passwordEnforced =
        store.getters.capabilities?.files_sharing?.public?.password?.enforced_for?.read_only ===
        true

      if (passwordEnforced) {
        return showQuickLinkPasswordModal(
          { store, $gettext: language.$gettext, passwordPolicyService },
          async (password) => {
            await createQuicklink({
              ability,
              clientService,
              language,
              password,
              resource: item,
              store
            })
          }
        )
      }

      await createQuicklink({
        ability,
        clientService,
        language,
        resource: item,
        store
      })
    },
    displayed: canShare
  }
} // FIXME: fix type, then add: satisfies ApplicationQuickActions
