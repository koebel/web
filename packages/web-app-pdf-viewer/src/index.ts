import translations from '../l10n/translations.json'
import { AppWrapperRoute } from 'web-pkg/src/components/AppTemplates/AppWrapperRoute'
import PdfViewer from './App.vue'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const routes = [
  {
    path: '/:driveAliasAndItem(.*)?',
    component: AppWrapperRoute(PdfViewer, {
      applicationId: 'pdf-viewer',
      urlForResourceOptions: {
        disposition: 'inline'
      }
    }),
    name: 'pdf-viewer',
    meta: {
      authContext: 'hybrid',
      title: $gettext('PDF Viewer'),
      patchCleanPath: true
    }
  }
]

const appInfo = {
  name: $gettext('PDF Viewer'),
  id: 'pdf-viewer',
  icon: 'resource-type-pdf',
  iconFillType: 'fill',
  iconColor: 'var(--oc-color-icon-pdf)',
  extensions: [
    {
      extension: 'pdf',
      routeName: 'pdf-viewer'
    }
  ]
}

export default {
  appInfo,
  routes,
  translations
}
