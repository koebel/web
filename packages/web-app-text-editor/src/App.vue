<template>
  <div
    :class="showPreview ? '' : 'oc-text-editor--no-preview'"
    class="oc-text-editor oc-width-1-1 oc-height-1-1"
  >
    <oc-textarea
      id="text-editor-input"
      :model-value="currentContent"
      name="input"
      label=""
      class="oc-height-1-1"
      :rows="20"
      :disabled="isReadOnly"
      @update:model-value="$emit('update:currentContent', $event)"
    />
    <!-- eslint-disable-next-line vue/no-v-html -->
    <div v-if="showPreview" id="text-editor-preview" v-html="renderedMarkdown" />
  </div>
</template>

<script lang="ts">
import { marked } from 'marked'
import sanitizeHtml from 'sanitize-html'
import { computed, defineComponent, unref, PropType } from 'vue'
import { Resource } from 'web-client/src/helpers/resource/types'
import { AppConfigObject } from 'web-pkg/src/apps/types'

export default defineComponent({
  name: 'TextEditor',
  props: {
    applicationConfig: { type: Object as PropType<AppConfigObject>, required: true },
    currentContent: {
      type: String,
      required: true
    },
    isReadOnly: { type: Boolean, required: false },
    resource: { type: Object as PropType<Resource>, required: true }
  },
  emits: ['update:currentContent'],
  setup(props) {
    const config = computed(() => {
      // TODO: Remove typecasting once vue-tsc has figured it out
      const { showPreviewOnlyMd = true } = props.applicationConfig as AppConfigObject
      return { showPreviewOnlyMd }
    })

    const showPreview = computed(() => {
      // FIXME: why is resource ever undefined/null?
      return props.resource?.extension === 'md' || !unref(config).showPreviewOnlyMd
    })

    const renderedMarkdown = computed(() => {
      return props.currentContent && unref(showPreview)
        ? sanitizeHtml(marked(props.currentContent))
        : null
    })

    return {
      showPreview,
      renderedMarkdown
    }
  }
})
</script>

<style lang="scss">
.oc-text-editor {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: 1fr 1fr;

  @media (min-width: $oc-breakpoint-medium-default) {
    grid-template-columns: 1fr 1fr;
    grid-template-rows: 1fr;
  }

  &--no-preview {
    grid-template-columns: 1fr;
    grid-template-rows: 1fr;
  }

  #text-editor-preview {
    max-height: 100%;
    overflow-y: auto;
    padding: var(--oc-space-xsmall) var(--oc-space-small);
    word-break: break-word;
  }

  #text-editor-input {
    resize: vertical;
  }
}
</style>
