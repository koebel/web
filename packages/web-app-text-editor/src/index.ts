import translations from '../l10n/translations.json'
import TextEditor from './App.vue'
import { AppWrapperRoute } from 'web-pkg/src/components/AppTemplates/AppWrapperRoute'

// just a dummy function to trick gettext tools
function $gettext(msg) {
  return msg
}

const appId = 'text-editor'

const routes = [
  {
    path: '/:driveAliasAndItem(.*)?',
    component: AppWrapperRoute(TextEditor, {
      applicationId: appId
    }),
    name: 'text-editor',
    meta: {
      authContext: 'hybrid',
      title: $gettext('Text Editor'),
      patchCleanPath: true
    }
  }
]

const fileExtensions = () => {
  const extensions: {
    extension: string
    label: string
    newFileMenu?: any
  }[] = [
    {
      extension: 'txt',
      label: $gettext('Plain text file')
    },
    {
      extension: 'md',
      label: $gettext('Markdown file')
    },
    {
      extension: 'js',
      label: $gettext('JavaScript file')
    },
    {
      extension: 'json',
      label: $gettext('JSON file')
    },
    {
      extension: 'xml',
      label: $gettext('XML file')
    },
    {
      extension: 'py',
      label: $gettext('Python file')
    },
    {
      extension: 'php',
      label: $gettext('PHP file')
    },
    {
      extension: 'yaml',
      label: $gettext('YAML file')
    }
  ]

  const config = (window as any).__$store.getters.extensionConfigByAppId(appId)
  extensions.push(...(config.extraExtensions || []).map((ext) => ({ extension: ext })))

  let primaryExtensions = (window as any).__$store.getters.extensionConfigByAppId(appId)
    .primaryExtensions || ['txt', 'md']

  if (typeof primaryExtensions === 'string') {
    primaryExtensions = [primaryExtensions]
  }

  return extensions.reduce((acc, extensionItem) => {
    const isPrimary = primaryExtensions.includes(extensionItem.extension)
    if (isPrimary) {
      extensionItem.newFileMenu = {
        menuTitle($gettext) {
          return $gettext(extensionItem.label)
        }
      }
    }
    acc.push(extensionItem)
    return acc
  }, [])
}

const appInfo = {
  name: $gettext('Text Editor'),
  id: appId,
  icon: 'file-text',
  isFileEditor: true,
  extensions: fileExtensions().map((extensionItem) => {
    return {
      extension: extensionItem.extension,
      ...(Object.prototype.hasOwnProperty.call(extensionItem, 'newFileMenu') && {
        newFileMenu: extensionItem.newFileMenu
      })
    }
  })
}

export default {
  appInfo,
  routes,
  translations
}
