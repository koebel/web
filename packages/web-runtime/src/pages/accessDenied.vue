<template>
  <div class="oc-height-viewport oc-flex oc-flex-column oc-flex-center oc-flex-middle">
    <div class="oc-login-card">
      <img class="oc-login-logo" :src="logoImg" alt="" :aria-hidden="true" />
      <div class="oc-login-card-body oc-width-medium">
        <h2 class="oc-login-card-title" v-text="cardTitle" />
        <p v-text="cardHint" />
        <oc-button
          v-if="accessDeniedHelpUrl"
          type="a"
          appearance="raw"
          :href="accessDeniedHelpUrl"
          target="_blank"
          ><span v-text="$gettext('Read more')"
        /></oc-button>
      </div>
      <div class="oc-login-card-footer oc-pt-rm">
        <p>
          {{ footerSlogan }}
        </p>
      </div>
    </div>
    <oc-button
      id="exitAnchor"
      class="oc-mt-m oc-width-medium"
      size="large"
      appearance="filled"
      variation="primary"
      v-bind="logoutButtonsAttrs"
      v-text="navigateToLoginText"
    />
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, unref } from 'vue'
import { queryItemAsString, useConfigurationManager, useRouteQuery, useStore } from 'web-pkg'
import { useGettext } from 'vue3-gettext'

export default defineComponent({
  name: 'AccessDeniedPage',
  setup() {
    const store = useStore()
    const configurationManager = useConfigurationManager()
    const redirectUrlQuery = useRouteQuery('redirectUrl')

    const { $gettext } = useGettext()

    const logoImg = computed(() => {
      return store.getters.configuration?.currentTheme?.logo?.login
    })

    const accessDeniedHelpUrl = computed(() => {
      return (
        store.getters.configuration?.commonTheme?.accessDeniedHelpUrl ||
        store.getters.configuration?.options?.accessDeniedHelpUrl
      )
    })
    const cardTitle = computed(() => {
      return $gettext('Not logged in')
    })
    const cardHint = computed(() => {
      return $gettext(
        'This could be because of a routine safety log out, or because your account is either inactive or not yet authorized for use. Please try logging in after a while or seek help from your Administrator.'
      )
    })
    const footerSlogan = computed(() => {
      return store.getters.configuration?.currentTheme?.general?.slogan
    })
    const navigateToLoginText = computed(() => {
      return $gettext('Log in again')
    })
    const logoutButtonsAttrs = computed(() => {
      const redirectUrl = queryItemAsString(unref(redirectUrlQuery))
      if (configurationManager.options.loginUrl) {
        const configLoginURL = new URL(encodeURI(configurationManager.options.loginUrl))
        if (redirectUrl) {
          configLoginURL.searchParams.append('redirectUrl', redirectUrl)
        }
        return {
          type: 'a',
          href: configLoginURL.toString()
        }
      }
      return {
        name: 'login',
        type: 'router-link',
        to: {
          name: 'login',
          query: {
            ...(redirectUrl && { redirectUrl })
          }
        }
      }
    })

    return {
      logoImg,
      cardTitle,
      cardHint,
      footerSlogan,
      navigateToLoginText,
      accessDeniedHelpUrl,
      logoutButtonsAttrs
    }
  }
})
</script>
