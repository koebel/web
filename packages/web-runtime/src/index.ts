import { DesignSystem as designSystem, pages, translations, supportedLanguages } from './defaults'
import { router } from './router'
import { configurationManager } from 'web-pkg/src/configuration'
import { createHead } from '@vueuse/head'
import { abilitiesPlugin } from '@casl/vue'
import { createMongoAbility } from '@casl/ability'
import merge from 'lodash-es/merge'

import {
  announceConfiguration,
  initializeApplications,
  announceApplicationsReady,
  announceClient,
  announceDefaults,
  announceClientService,
  announceStore,
  announceTheme,
  announceCustomStyles,
  announceTranslations,
  announceVersions,
  announceUppyService,
  announceAuthService,
  startSentry,
  announceCustomScripts,
  announceLoadingService,
  announcePreviewService,
  announcePasswordPolicyService,
  announceAdditionalTranslations
} from './container/bootstrap'
import { applicationStore } from './container/store'
import {
  buildPublicSpaceResource,
  buildSpace,
  isPersonalSpaceResource,
  isPublicSpaceResource,
  Resource
} from 'web-client/src/helpers'
import { loadCustomTranslations } from 'web-runtime/src/helpers/customTranslations'
import { WebDAV } from 'web-client/src/webdav'
import { DavProperty } from 'web-client/src/webdav/constants'
import { computed, createApp } from 'vue'
import PortalVue, { createWormhole } from 'portal-vue'
import { createPinia } from 'pinia'
import Avatar from './components/Avatar.vue'
import focusMixin from './mixins/focusMixin'
import { ArchiverService } from 'web-pkg/src/services/archiver'
import { get } from 'lodash-es'

export const bootstrapApp = async (configurationPath: string): Promise<void> => {
  const pinia = createPinia()
  const app = createApp(pages.success)
  app.use(pinia)

  app.provide('$router', router)

  const runtimeConfiguration = await announceConfiguration(configurationPath)
  startSentry(runtimeConfiguration, app)

  const store = await announceStore({ runtimeConfiguration })
  app.provide('$store', store)
  app.provide('store', store)

  app.use(abilitiesPlugin, createMongoAbility([]), { useGlobalProperties: true })

  const gettext = announceTranslations({
    app,
    availableLanguages: supportedLanguages,
    translations
  })
  announceUppyService({ app })

  announceClientService({ app, runtimeConfiguration, configurationManager, store })
  // TODO: move to announceArchiverService function
  app.config.globalProperties.$archiverService = new ArchiverService(
    app.config.globalProperties.$clientService,
    store.getters.configuration.server || window.location.origin,
    computed(() =>
      get(store, 'getters.capabilities.files.archivers', [
        {
          enabled: true,
          version: '1.0.0',
          formats: ['tar', 'zip'],
          archiver_url: `${store.getters.configuration.server}index.php/apps/files/ajax/download.php`
        }
      ])
    )
  )
  app.provide('$archiverService', app.config.globalProperties.$archiverService)
  announceLoadingService({ app })
  announcePreviewService({ app, store, configurationManager })
  await announceClient(runtimeConfiguration)

  const applicationsPromise = initializeApplications({
    app,
    runtimeConfiguration,
    configurationManager,
    store,
    supportedLanguages,
    router,
    gettext
  })

  const customTranslationsPromise = loadCustomTranslations({ configurationManager })
  const themePromise = announceTheme({ store, app, designSystem, runtimeConfiguration })
  const [customTranslations] = await Promise.all([
    customTranslationsPromise,
    applicationsPromise,
    themePromise
  ])

  announceAdditionalTranslations({ gettext, translations: merge(customTranslations) })

  announceAuthService({ app, configurationManager, store, router })
  announceCustomStyles({ runtimeConfiguration })
  announceCustomScripts({ runtimeConfiguration })
  announceDefaults({ store, router })

  app.use(router)
  app.use(store)
  app.use(createHead())
  app.config.globalProperties.$wormhole = createWormhole()
  app.use(PortalVue, {
    wormhole: app.config.globalProperties.$wormhole
  })

  app.component('AvatarImage', Avatar)
  app.mixin(focusMixin)

  app.mount('#owncloud')

  const applications = Array.from(applicationStore.values())
  applications.forEach((application) => application.mounted(app))

  store.watch(
    (state, getters) =>
      getters['runtime/auth/isUserContextReady'] ||
      getters['runtime/auth/isIdpContextReady'] ||
      getters['runtime/auth/isPublicLinkContextReady'],
    async (newValue, oldValue) => {
      if (!newValue || newValue === oldValue) {
        return
      }
      announceVersions({ store })
      await announceApplicationsReady({ app, store, applications })
    },
    {
      immediate: true
    }
  )

  store.watch(
    (state, getters) => {
      return getters['runtime/auth/isUserContextReady']
    },
    async (userContextReady) => {
      if (!userContextReady) {
        return
      }

      announcePasswordPolicyService({ app, store })

      const clientService = app.config.globalProperties.$clientService

      // Load spaces to make them available across the application
      if (store.getters.capabilities?.spaces?.enabled) {
        const graphClient = clientService.graphAuthenticated
        await store.dispatch('runtime/spaces/loadSpaces', { graphClient })
        const personalSpace = store.getters['runtime/spaces/spaces'].find((space) =>
          isPersonalSpaceResource(space)
        )

        if (!personalSpace) {
          return
        }

        store.commit('runtime/spaces/UPDATE_SPACE_FIELD', {
          id: personalSpace.id,
          field: 'name',
          value: app.config.globalProperties.$gettext('Personal')
        })
        return
      }

      // Spaces feature not available. Create a virtual personal space
      const user = store.getters.user
      const space = buildSpace({
        id: user.id,
        driveAlias: `personal/${user.id}`,
        driveType: 'personal',
        name: app.config.globalProperties.$gettext('All files'),
        webDavPath: `/files/${user.id}`,
        webDavTrashPath: `/trash-bin/${user.id}`,
        serverUrl: configurationManager.serverUrl
      })
      const personalHomeInfo = await (clientService.webdav as WebDAV).getFileInfo(
        space,
        {
          path: ''
        },
        { davProperties: [DavProperty.FileId] }
      )
      space.fileId = personalHomeInfo.fileId
      store.commit('runtime/spaces/ADD_SPACES', [space])
      store.commit('runtime/spaces/SET_SPACES_INITIALIZED', true)
    },
    {
      immediate: true
    }
  )
  store.watch(
    (state, getters) => {
      return getters['runtime/auth/isPublicLinkContextReady']
    },
    (publicLinkContextReady) => {
      if (!publicLinkContextReady) {
        return
      }
      // Create virtual space for public link
      const publicLinkToken = store.getters['runtime/auth/publicLinkToken']
      const publicLinkPassword = store.getters['runtime/auth/publicLinkPassword']
      const space = buildPublicSpaceResource({
        id: publicLinkToken,
        name: app.config.globalProperties.$gettext('Public files'),
        ...(publicLinkPassword && { publicLinkPassword }),
        serverUrl: configurationManager.serverUrl
      })
      store.commit('runtime/spaces/ADD_SPACES', [space])
      store.commit('runtime/spaces/SET_SPACES_INITIALIZED', true)
    },
    {
      immediate: true
    }
  )
  store.watch(
    // only needed if a public link gets re-resolved with a changed password prop (changed or removed).
    // don't need to set { immediate: true } on the watcher.
    (state, getters) => {
      return getters['runtime/auth/publicLinkPassword']
    },
    (publicLinkPassword: string | undefined) => {
      const publicLinkToken = store.getters['runtime/auth/publicLinkToken']
      const space = store.getters['runtime/spaces/spaces'].find((space: Resource) => {
        return isPublicSpaceResource(space) && space.id === publicLinkToken
      })
      if (!space) {
        return
      }
      space.publicLinkPassword = publicLinkPassword
    }
  )
}

export const bootstrapErrorApp = async (err: Error): Promise<void> => {
  const store = await announceStore({ runtimeConfiguration: {} })
  announceVersions({ store })
  const app = createApp(pages.failure)
  await announceTheme({ store, app, designSystem })
  console.error(err)
  app.use(store)
  await announceTranslations({
    app,
    availableLanguages: supportedLanguages,
    translations
  })
  app.mount('#owncloud')
}
;(window as any).runtimeLoaded({
  bootstrapApp,
  bootstrapErrorApp
})
