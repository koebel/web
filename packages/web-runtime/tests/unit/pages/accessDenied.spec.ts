import accessDenied from '../../../src/pages/accessDenied.vue'
import {
  createStore,
  defaultComponentMocks,
  defaultPlugins,
  mount,
  defaultStoreMockOptions
} from 'web-test-helpers'
import { mock } from 'jest-mock-extended'

const selectors = {
  logInAgainButton: '#exitAnchor'
}

import { ConfigurationManager, useConfigurationManager } from 'web-pkg/src'

jest.mock('web-pkg/src/composables/configuration/useConfigurationManager')

describe('access denied page', () => {
  it('renders component', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  describe('"Log in again" button', () => {
    it('navigates to "loginUrl" if set in config', () => {
      const loginUrl = 'https://myidp.int/login'
      const { wrapper } = getWrapper({ loginUrl })

      const logInAgainButton = wrapper.find(selectors.logInAgainButton)
      const loginAgainUrl = new URL(logInAgainButton.attributes().href)
      loginAgainUrl.search = ''

      expect(logInAgainButton.exists()).toBeTruthy()
      expect(loginAgainUrl.toString()).toEqual(loginUrl)
    })
  })
})

function getWrapper({ loginUrl = '' } = {}) {
  const mocks = {
    ...defaultComponentMocks()
  }
  const storeOptions = { ...defaultStoreMockOptions }

  jest.mocked(useConfigurationManager).mockImplementation(() =>
    mock<ConfigurationManager>({
      options: {
        loginUrl
      }
    } as any)
  )

  const store = createStore(storeOptions)

  return {
    storeOptions,
    mocks,
    wrapper: mount(accessDenied, {
      global: {
        plugins: [...defaultPlugins(), store],
        mocks,
        provide: mocks
      }
    })
  }
}
