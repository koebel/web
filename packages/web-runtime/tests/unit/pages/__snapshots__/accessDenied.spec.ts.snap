// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`access denied page renders component 1`] = `
<div class="oc-height-viewport oc-flex oc-flex-column oc-flex-center oc-flex-middle">
  <div class="oc-login-card">
    <img alt="" aria-hidden="true" class="oc-login-logo">
    <div class="oc-login-card-body oc-width-medium">
      <h2 class="oc-login-card-title">Not logged in</h2>
      <p>This could be because of a routine safety log out, or because your account is either inactive or not yet authorized for use. Please try logging in after a while or seek help from your Administrator.</p>
      <!--v-if-->
    </div>
    <div class="oc-login-card-footer oc-pt-rm">
      <p></p>
    </div>
  </div>
  <a attrs="[object Object]" class="oc-button oc-rounded oc-button-l oc-button-justify-content-center oc-button-gap-m oc-button-primary oc-button-primary-filled oc-mt-m oc-width-medium" id="exitAnchor" name="login">Log in again</a>
</div>
`;
